pulp_docker/app/tasks/stages/dedupe_save.py

```diff
@@ -0,0 +1,163 @@
+from django.db.models import Q
+from pulpcore.plugin.stages import Stage
+from pulpcore.plugin.models import Artifact, ContentArtifact, RemoteArtifact
+# from pulpcore.plugin.models import ProgressBar
+
+import logging
+log = logging.getLogger("STUPIDSAVE")
+
+
+class SerialArtifactSave(Stage):
+    """
+    Save Artifacts one at a time, combining duplicates.
+    """
+
+    async def __call__(self, in_q, out_q):
+        """
+        The coroutine for this stage.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): The queue to receive
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+            out_q (:class:`asyncio.Queue`): The queue to put
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+        Returns:
+            The coroutine for this stage.
+
+        """
+        while True:
+            dc = await in_q.get()
+            if dc is None:
+                break
+            self.query_and_save_artifacts(dc)
+            await out_q.put(dc)
+        await out_q.put(None)
+
+    def query_and_save_artifacts(self, dc):
+        """
+        Combine duplicated artifacts, save unique artifacts.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): The queue to receive
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+        """
+        for da in dc.d_artifacts:
+            artifact_q = Q()
+            for digest_name in da.artifact.DIGEST_FIELDS:
+                digest_value = getattr(da.artifact, digest_name)
+                if digest_value:
+                    key = {digest_name: digest_value}
+                    artifact_q &= Q(**key)
+            try:
```

Comment:
This _query-and-save_ logic is vulnerable to race conditions.  A better design pattern is:
```
try:
    model.save()
except IntegrityError:  # dup
    model = model.objects.get(**natural_key)
```

Unfortunately, django does not have an explicit duplicate key exception.

Reply:
I'm fine with this change, its more pythonic anyway.

However, AFAIK it is not vulnerable to race conditions given my understanding of coroutines. Since coroutines don't actually run multiple processes/threads in parallel and this is not an async function, my assumption is that this entire block will execute all at once without any other coroutines executing during that time. If this is not correct, please let me know.